### DISCUSSION TITLE
Use Bencher to benchmark PRs and track metrics

### DISCUSSION THREAD

**Discussion by epompeii** [diesel-rs/diesel#3848] [Ideas]
Would the `diesel` project be interested in a PR to start tracking benchmarks and catching performance regressions with [Bencher](https://github.com/bencherdev/bencher)?

I ran across @weiznich post on Reddit regarding diesel benchmarking: https://old.reddit.com/r/rust/comments/15h6rsv/crate_authors_how_do_you_prevent_performance/juqk4ft/

Is there anything that https://github.com/diesel-rs/metrics does not provide that you would like to have in a benchmark tracking tool?

**Comment by weiznich** [comment]
Can you explain what Bencher offers what's not covered by the current setup?

**Reply by epompeii** [reply]
Sure!

> I did not find a way to post the results to the PR itself, especially if the PR is coming from a fork. That means you need to inspect the output of the relevant CI run for the numbers.

Bencher [supports posting on PRs](https://github.com/bencherdev/bencher#comment-on-prs) and with a bit of setup it works on PRs from forks as well: https://bencher.dev/docs/how-to/github-actions/#pull-requests

Currently it appears you're doing pairwise benchmarking, running the `master` and PR branch benchmarks and then comparing the results. Bencher would open up the possibility of tracking the historical metrics and using those plus statistical thresholds to detect regressions in PRs. This could be used to fail jobs/PRs if there is severe performance regression. This would also obviate the need to separately run the historical benchmarks on `master`.

If you would prefer to keep with the pairwise + separate run on `master approach, you could still use Bencher for detecting performance regressions in both cases and failing jobs/PRs.

There is also the plus of being able to dynamically inspect and plot the data using the Bencher UI.

**Reply by epompeii** [reply]
Also, thank you for your work on Diesel!
I use it in the Bencher API server 😃

**Reply by weiznich** [reply]
Posting to PR's would be helpful, although I've mostly solved that by using action summaries. See [this PR](https://github.com/diesel-rs/diesel/pull/3809) for an example. It's currently by design that we run benchmarks only for certain PR by manually applying labels. That shouldn't be changed due to having a limited amount of CI workers.

As for tracking data over time. We have a different job for that, that runs more benchmarks (also for other crates). It's important for me to have the resulting data open, therefore they are pushed to a git repository. In addition: From the examples on the bencher webpage it seems like you offer only a graph of results over time. It's nice to have that as interactive graph, but we are also interested in comparison graphs for the last run, like that one ![image](https://github.com/diesel-rs/diesel/assets/1674512/3f9fecb5-5da0-4aaf-9c5c-56d9276b425a) (Although that would be more useful if it was interactive as well). 

To summarize what is written above: I feel that there are still features missing to move the benchmark over to bencher. Most notably:

* The ability to run the benchmarks only for certain PR based on tags
* Advanced analytic formats (boxcharts to compare values for a specific run + maybe data tables that show [min, max, media, mean] per group + diff to a reference value)

**Reply by epompeii** [reply]
Okay, that all makes sense.

> The ability to run the benchmarks only for certain PR based on tags

I believe that this can be accomplished with Bencher in the same way you all do it now. The main change would be here:

https://github.com/diesel-rs/diesel/blob/b51567f6f4bf32d03c3c00b03f38b0a3c29899f0/.github/workflows/benches.yml#L54

To using Bencher, something like this:

```yml
 run: bencher run --err 'cargo +stable bench --manifest-path diesel_bench/Cargo.toml --no-default-features --features "${{matrix.backend}}"'
```

You wouldn't need the following checkout and `critcmp`.

> Advanced analytic formats (boxcharts to compare values for a specific run + maybe data tables that show [min, max, media, mean] per group + diff to a reference value)

This I totally get! I have a tracking issue for point-in-time comparisons (https://github.com/bencherdev/bencher/issues/72).
In the mean time, would you be interested in a PR that also starts saving the metrics to Bencher in addition to https://github.com/diesel-rs/metrics ?
[This is what `rustls` is about to start doing](https://github.com/aochagavia/rustls-bench-app/pull/6). The change would be somewhere around here:
https://github.com/diesel-rs/diesel/blob/b51567f6f4bf32d03c3c00b03f38b0a3c29899f0/.github/workflows/metrics.yml#L78

Then you can have the best of both worlds while I work on adding new plots/tables.

> It's important for me to have the resulting data open

For a public project like this one, anyone can query the API endpoint for the data on Bencher. No auth or signup required.
[This is all the metrics for the past month of Bencher dogfooding its `devel` branch](https://api.bencher.dev/v0/projects/bencher/perf?start_time=1696784565000&end_time=1699376668000&report=1e7a3e79-68d0-4d07-a903-bb5c75429fa6&metric_kind=latency&branches=e2075f16-da03-4229-9b46-5b96bb00c77e&testbeds=0d991aac-b241-493a-8b0f-8d41419455d2&benchmarks=e93b3d71-8499-4fae-bb7c-4e540b775714%2C3bfd5887-83ec-4e62-8690-02855a38fbc9%2C3525f177-fc8f-4a92-bd2f-dda7c4e15699%2C5655ed2a-3e45-4622-bdbd-39cdd9837af8%2C1db23e93-f909-40aa-bf42-838cc7ae05f5)
I'm curious if you would like things to be anymore open than that?

**Reply by weiznich** [reply]
> This I totally get! I have a tracking issue for point-in-time comparisons (https://github.com/bencherdev/bencher/issues/72).
In the mean time, would you be interested in a PR that also starts saving the metrics to Bencher in addition to https://github.com/diesel-rs/metrics ?
https://github.com/aochagavia/rustls-bench-app/pull/6. The change would be somewhere around here:

That issue does not completely cover what I would like to see as bar charts are different to the box plots that are currently used. Box plots cover the statistical distribution for each set of values, while a bar chart merely covers a single value per group.

> For a public project like this one, anyone can query the API endpoint for the data on Bencher. No auth or signup required.
[This is all the metrics for the past month of Bencher dogfooding its devel branch](https://api.bencher.dev/v0/projects/bencher/perf?start_time=1696784565000&end_time=1699376668000&report=1e7a3e79-68d0-4d07-a903-bb5c75429fa6&metric_kind=latency&branches=e2075f16-da03-4229-9b46-5b96bb00c77e&testbeds=0d991aac-b241-493a-8b0f-8d41419455d2&benchmarks=e93b3d71-8499-4fae-bb7c-4e540b775714%2C3bfd5887-83ec-4e62-8690-02855a38fbc9%2C3525f177-fc8f-4a92-bd2f-dda7c4e15699%2C5655ed2a-3e45-4622-bdbd-39cdd9837af8%2C1db23e93-f909-40aa-bf42-838cc7ae05f5)
I'm curious if you would like things to be anymore open than that?

I'm sorry, but I trust github more to not go away anytime soon than a new service like bencher. That's the reason I prefer to have the data just there in the repo. (Also migrating to bencher would loose the old data, right?) If it's possible to do both I'm open to go that way. 


I think I would be open with having bencher as additional target for the results and keep both solutions in parallel for some time. After that we can see how well each solution works and decide how we want to continue there.

**Reply by epompeii** [reply]
> That issue does not completely cover what I would like to see as bar charts are different to the box plots that are currently used. 

Very true! I have updated the issue to include box plots as a part of it. I see that working in a similar fashion to the toggling of the x-axis in the current line plots (datetime <-> branch version), so it would toggle between: bar chart <-> box plot. Thank you for making sure I note the difference!

> I'm sorry, but I trust github more to not go away anytime soon than a new service like bencher

Totally fair! I'd feel the same way.

> Also migrating to bencher would loose the old data, right?

Bencher can import the old data. It's possible to backdate reports. I think we should hold off on that for now though.

> I think I would be open with having bencher as additional target for the results and keep both solutions in parallel for some time. After that we can see how well each solution works and decide how we want to continue there.

Sounds good! I'll start working on a PR for tracking the `master` branch metrics.

**Reply by epompeii** [reply]
PR is up: https://github.com/diesel-rs/diesel/pull/3849

Please, let me know what you think! 😃
